feat(websocket-websys): add support for different WASM environments

jsdanielh opened the pull request:
## Description

We introduce a `WebContext` `enum` that abstracts and detects the `Window` vs the `WorkerGlobalScope` API.

Related: https://github.com/rustwasm/wasm-bindgen/issues/1046.

## Change checklist

<!-- Please add a Changelog entry in the appropriate crates and bump the crate versions if needed. See <https://github.com/libp2p/rust-libp2p/blob/master/docs/release.md#development-between-releases>-->

- [X] I have performed a self-review of my own code
- [X] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] A changelog entry has been made in the appropriate crates

sisou reviewed (approved):
For some added context: this uses the same approach as for gloo-timers: https://github.com/rustwasm/gloo/pull/185, extended already by the codegen fix of https://github.com/rustwasm/gloo/pull/283.

This makes it possible to run this transport in environments that support `setInterval`, but don't have a `window` global. This way it also works in NodeJS, Deno and Bun (although Deno provides a `window` for compatibility).

thomaseizinger commented on transports/websocket-websys/CHANGELOG.md:
```diff
@@ -1,3 +1,8 @@
+## 0.3.1
+
+- Add support for workers by binding to the global `setInterval` and `clearInterval`
+  functions directly.
```
```suggestion
- Add support for workers by binding to the global `setInterval` and `clearInterval` functions directly.
```

thomaseizinger replied:
Needs PR reference please.

jsdanielh replied:
Done

thomaseizinger reviewed (commented):
Thanks!

I am not entirely opposed to this change but it isn't particularly clean. The idea of `libp2p-websocket-websys` was that we provide WASM bindings **via web-sys**. 

We are now working around the `web-sys` crate which is a bit odd. Would it make sense to perhaps upstream this change to `web-sys` instead by offering `setInterval` and `clearInterval` as top-level functions instead of always going through `window`?

jsdanielh committed:
fix(websocket-websys): Add support for different WASM environments

Add support different WASM environments (such as workers, NodeJS)
that don't have a `window` global. This is done by introducing a
`WebContext` `enum` that abstracts and detects the `Window` vs the
`WorkerGlobalScope` API.
This is done due to the `web-sys` lack of interest to support this
(see discussion in
[this issue](https://github.com/rustwasm/wasm-bindgen/issues/1046)).

thomaseizinger commented on transports/websocket-websys/CHANGELOG.md:
```diff
@@ -1,3 +1,10 @@
+- Add support for different WASM environments by introducing a `WebContext` that
+  detects and abstracts the `Window` vs the `WorkerGlobalScope` API. See [PR 4889].
+
+[PR 4889]: https://github.com/libp2p/rust-libp2p/pull/4889
```
```suggestion
- Add support for different WASM environments by introducing a `WebContext` that detects and abstracts the `Window` vs the `WorkerGlobalScope` API.
  See [PR 4889](https://github.com/libp2p/rust-libp2p/pull/4889).
```

thomaseizinger commented on transports/websocket-websys/src/web_context.rs:
```diff
@@ -0,0 +1,62 @@
+        match self {
+            WebContext::Window(w) => w.set_interval_with_callback_and_timeout_and_arguments(
+                handler,
+                timeout, // Chosen arbitrarily and likely worth tuning. Due to low impact of the /ws transport, no further effort was invested at the time.
```
I don't think this comment makes any sense here because we are just passing on the variable.

thomaseizinger commented on transports/websocket-websys/src/web_context.rs:
```diff
@@ -0,0 +1,62 @@
+        } else if !global.worker().is_undefined() {
+            Self::Worker(global.unchecked_into())
+        } else {
+            panic!("Only supported in a browser or web worker");
```
I think I'd prefer `new` to return an `Option` and leave the panicking to upper layers.

thomaseizinger commented on transports/websocket-websys/src/web_context.rs:
```diff
@@ -0,0 +1,62 @@
+
+        let global: Global = js_sys::global().unchecked_into();
+
+        if !global.window().is_undefined() {
```
Would it make sense to use the global `window` function here? That would be one less method we need to bind to manually.

It is unfortunate that `web_sys` doesn't offer a global `worker` function.

thomaseizinger commented on transports/websocket-websys/src/lib.rs:
```diff
@@ -35,7 +37,17 @@ use std::sync::atomic::{AtomicBool, Ordering};
+thread_local! {
+    /// Global web context for abstracting the `Window` vs the `WorkerGlobalScope` API
+    static GLOBAL_WEB_CONTEXT: WebContext = WebContext::new();
+}
```
Why do we need the thread-local? We only call this in the constructor and `Drop` impl so I don't think we'd be able to measure a performance benefit here.

thomaseizinger commented on transports/websocket-websys/src/lib.rs:
```diff
@@ -35,7 +37,17 @@ use std::sync::atomic::{AtomicBool, Ordering};
+use crate::web_context::WebContext;
+
+/// Arbitrary, maximum amount we are willing to buffer before we throttle our user.
+const MAX_BUFFER: usize = 1024 * 1024;
```
Can we move this down again to reduce the diff to a minimum? Thanks!

thomaseizinger reviewed (commented):
Thank you!

Overall concept looks great, just a few nits :)

thomaseizinger reviewed (approved):
Great work!

Thank you for iterating on this with me :)

mergify[bot] committed:
Merge branch 'master' into pr
